pymodbus/datastore/simulator.py

```diff
@@ -724,8 +724,11 @@ def validate_type(self, func_code, real_address, count):
         for i in range(  # pylint: disable=consider-using-any-or-all
             real_address, real_address + count, reg_step
         ):
-            if self.registers[i].type not in check:
-                return False
+            if self.registers[i].type in check:
+                continue
+            if self.registers[i].type is CellType.NEXT:
```

Comment:
This is not right: real_address + 0 may not be NEXT, because it means you read off base.

For BITS, UINT16 next is never allowed (they occupy exact 1 register).
For UINT32, FLOAT32, first address must never be NEXT, second address must be NEXT, 3 address never etc....
For STRING first address must be STRING, next addresses must be NEXT / STRING.

So it is a bit more complex.

Reply:
As I said before, I don't really understand the NEXT celltype.
Why should it not be authorized to read the UINT32 in two requests or to read two UINT32 values in one request, but it's authorized to read only one register of the UINT32.
For strings it's similar, if there is a longer string and I would like to get only a part of it, why this wouldn't be allowed?
Initially I thought the type check is more of a function code check.

Reply:
NEXT is quite easy, it is the second register in UINT32, FLOAT32, and remaining registers in STRING.

Consider this sequence of registers:
12: UINT32,  13: NEXT,  14: UINT32, 15: NEXT, 16:UINT32, 17:NEXT

The following are valid reads:
Read_holding(12, 2)
Read_holding(12, 6)
Read_holding(14,2)

The following are invalid (Because they do not read whole UINT32)
Read_holding(12, 1)
Read_holding(13, 1)
Read_holding(12, 3)

FLOAT32 are exactly the same.

STRINGS are not so different:
12: STRING, 13: NEXT, 14: NEXT, 15: NEXT, 16: STRING, 17: NEXT

The following are valid:
read_holding(12, 4)
read_holding(12, 6) # read 2 complete strings (allowed just as it is allowed to read multiple UINT32)
read_holding(16, 2)

Type check is there to ensure you read the strings correctly, so the following is not allowed:
read_holding(12, 1)
read_holding(12, 1)
read_holding(13, 1)

Does that help ?

Reply:
The type check is there to ensure you read the registers as the designer of the device intended. It is a "verification", and you can find X reasons why it should be different, but then the answer is configure it to False!

Reply:
yes, much better :)